File: doordash/src/main/kotlin/com/quantum/doordash/config/ExternalConfig.kt
package com.quantum.doordash.config

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration

private val mapper = jacksonObjectMapper()
private val httpClient: HttpClient = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(5)).build()

fun applyExternalConfig(config: AppConfig): AppConfig {
    var resolved = config

    if (config.configFromConsul) {
        try {
            val payload = fetchConsulConfig(config)
            if (payload != null) {
                resolved = resolved.copy(
                    httpPort = payload["http_port"]?.toString()?.toIntOrNull() ?: resolved.httpPort,
                    dbUrl = payload["db_url"]?.toString() ?: resolved.dbUrl,
                    dbUser = payload["db_user"]?.toString() ?: resolved.dbUser,
                    dbPassword = payload["db_password"]?.toString() ?: resolved.dbPassword,
                    dbSchema = payload["db_schema"]?.toString() ?: resolved.dbSchema,
                    redisUrl = payload["redis_url"]?.toString() ?: resolved.redisUrl,
                    temporalAddress = payload["temporal_address"]?.toString() ?: resolved.temporalAddress,
                    temporalNamespace = payload["temporal_namespace"]?.toString() ?: resolved.temporalNamespace,
                    taskQueue = payload["task_queue"]?.toString() ?: resolved.taskQueue,
                    apiKey = payload["api_key"]?.toString() ?: resolved.apiKey,
                )
            }
        } catch (exc: Exception) {
            println("consul config fetch failed: ${exc.message}")
        }
    }

    if (config.secretsFromVault) {
        try {
            val secretData = fetchVaultSecret(config)
            val apiKey = secretData?.get(config.vaultApiKeyField)?.asText()
            if (!apiKey.isNullOrBlank()) {
                resolved = resolved.copy(apiKey = apiKey)
            }
        } catch (exc: Exception) {
            println("vault secret fetch failed: ${exc.message}")
        }
    }

    return resolved
}

private fun fetchConsulConfig(config: AppConfig): Map<String, Any?>? {
    val keyPath = config.consulKvPath.trimStart('/')
    val url = "${config.consulHttpAddr.trimEnd('/')}/v1/kv/$keyPath?raw"

    val request = HttpRequest.newBuilder()
        .uri(URI.create(url))
        .timeout(Duration.ofSeconds(5))
        .GET()
        .build()

    val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
    if (response.statusCode() == 404) {
        return null
    }
    if (response.statusCode() >= 400) {
        throw RuntimeException("consul status ${response.statusCode()}: ${response.body()}")
    }
    return mapper.readValue(response.body(), Map::class.java) as Map<String, Any?>
}

private fun fetchVaultSecret(config: AppConfig): com.fasterxml.jackson.databind.JsonNode? {
    val path = config.vaultSecretPath.trimStart('/')
    val url = "${config.vaultAddr.trimEnd('/')}/v1/${config.vaultKvMount}/data/$path"

    val request = HttpRequest.newBuilder()
        .uri(URI.create(url))
        .header("X-Vault-Token", config.vaultToken)
        .timeout(Duration.ofSeconds(5))
        .GET()
        .build()

    val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
    if (response.statusCode() == 404) {
        return null
    }
    if (response.statusCode() >= 400) {
        throw RuntimeException("vault status ${response.statusCode()}: ${response.body()}")
    }

    val root = mapper.readTree(response.body())
    return root.path("data").path("data")
}

File: algorand/src/algorand/core/external_config.py
"""Optional external configuration/secrets from Consul and Vault."""

from __future__ import annotations

import logging
from typing import Any

import httpx

logger = logging.getLogger(__name__)


def apply_external_settings(settings):
    updates: dict[str, Any] = {}

    if settings.config_from_consul:
        try:
            updates.update(_fetch_consul_config(settings))
        except Exception as exc:
            logger.warning("Consul config fetch failed: %s", exc)

    if settings.secrets_from_vault:
        try:
            secret_data = _fetch_vault_secret(settings)
            if settings.vault_api_key_field in secret_data:
                updates["api_key"] = secret_data[settings.vault_api_key_field]
        except Exception as exc:
            logger.warning("Vault secret fetch failed: %s", exc)

    if not updates:
        return settings

    allowed = set(settings.model_fields.keys())
    filtered_updates = {key: value for key, value in updates.items() if key in allowed}
    if not filtered_updates:
        return settings

    merged = {**settings.model_dump(), **filtered_updates}
    return settings.__class__.model_validate(merged)


def _fetch_consul_config(settings) -> dict[str, Any]:
    key_path = settings.consul_kv_path.lstrip("/")
    url = f"{settings.consul_http_addr.rstrip('/')}/v1/kv/{key_path}?raw"
    response = httpx.get(url, timeout=5.0)
    if response.status_code == 404:
        return {}
    response.raise_for_status()
    payload = response.json()
    if not isinstance(payload, dict):
        raise RuntimeError("consul config payload must be a JSON object")
    return payload


def _fetch_vault_secret(settings) -> dict[str, Any]:
    secret_path = settings.vault_secret_path.lstrip("/")
    url = f"{settings.vault_addr.rstrip('/')}/v1/{settings.vault_kv_mount}/data/{secret_path}"
    response = httpx.get(
        url,
        headers={"X-Vault-Token": settings.vault_token},
        timeout=5.0,
    )
    if response.status_code == 404:
        return {}
    response.raise_for_status()
    payload = response.json()
    secret_data = payload.get("data", {}).get("data", {})
    if not isinstance(secret_data, dict):
        return {}
    return secret_data
